### PR TITLE
add distillation

### PR DESCRIPTION
**Pull request by LDOUBLEV**
add distillation

* new features
    * add new arch to support distillation
    * add new losses and post process
* improvements
    * remove names in the models cause they are not used in the dygraph mode
* todo
- [x] model export/save for distillation models
- [x] remove `prertain` and `freeze` paramster out of the distillation arch
- [x] change basemodel output from tensor to dict
- [x] support for multiple metrics' info output and logger
- [x] polish load pretrain

### CODE REVIEW

File: configs/rec/ch_ppocr_v2.0/rec_chinese_lite_train_distillation_v2.1.yml [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+    name: Cosine
+    learning_rate: 0.0005
+    warmup_epoch: 5
+  regularizer:
+    name: L2
+    factor: 1.0e-05
+Architecture:
+  name: DistillationModel
+  algorithm: Distillation
+  freeze_params:
```
Comment (LDOUBLEV) [inline configs/rec/ch_ppocr_v2.0/rec_chinese_lite_train_distillation_v2.1.yml:40]: freeze_params, pretrained 可以作为Student 的一个属性

这里的pretrained 优先级低于Global.pretrain_model，而且student.pretrain隐藏的太深了，没有暴露在main函数中
另外可能会踩坑，Global.pretrained_model加载模型覆盖了Student.pretrained的模型，实际上既然设置了student.pretrain，希望是从这里加载模型，不是被Global覆盖掉

File: configs/rec/ch_ppocr_v2.0/rec_chinese_lite_train_distillation_v2.1.yml [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+  - DistillationCTCLoss:
+      weight: 1.0
+      model_name_list: ["Student", "Teacher"]
+      key: null
+  - DistillationDMLLoss:
+      weight: 1.0
+      model_name_list1: ["Student"]
+      model_name_list2: ["Teacher"]
+
+PostProcess:
```
Comment (LDOUBLEV) [inline configs/rec/ch_ppocr_v2.0/rec_chinese_lite_train_distillation_v2.1.yml:91]: 后处理单独抽出来存在一个问题是，eval的时候，实际上得到了student和teacher的输出，但是评估的时候是通过key_out指定计算哪些网络后处理去计算eval，假设是student和teacher 都进行后处理，这样得到的有两个后处理结果，现有的评估函数不支持dict结果传入，那还得重新写一个eval函数？

File: configs/rec/ch_ppocr_v2.0/rec_chinese_lite_train_distillation_v2.1.yml [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+Loss:
+  name: CombinedLoss
+  loss_config_list:
+  - DistillationCTCLoss:
+      weight: 1.0
+      model_name_list: ["Student", "Teacher"]
+      key: null
+  - DistillationDMLLoss:
+      weight: 1.0
+      model_name_list1: ["Student"]
```
Comment (LDOUBLEV) [inline configs/rec/ch_ppocr_v2.0/rec_chinese_lite_train_distillation_v2.1.yml:88]: model_name_list 1 2 名字不够 望文知意，可以考虑修改；

File: configs/rec/ch_ppocr_v2.0/rec_chinese_lite_train_distillation_v2.1.yml [hunk 4]
```diff
@@ -0,0 +1,151 @@
+Global:
+  debug: false
+  use_gpu: true
+  epoch_num: 800
+  log_smooth_window: 20
+  print_batch_step: 10
+  save_model_dir: ./output/rec_D081
+  save_epoch_step: 3
+  eval_batch_step:
+  - 0
```
Comment (LDOUBLEV) [inline configs/rec/ch_ppocr_v2.0/rec_chinese_lite_train_distillation_v2.1.yml:10]: 这里改成list形式，和其他yml文件格式保持一致

File: ppocr/losses/basic_loss.py [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+            one_hot_target = target
+        soft_target = F.label_smooth(one_hot_target, epsilon=self.epsilon)
+        soft_target = paddle.reshape(soft_target, shape=[-1, class_num])
+        return soft_target
+
+    def forward(self, x, label):
+        loss_dict = {}
+        if self.epsilon is not None:
+            class_num = x.shape[-1]
+            label = self._labelsmoothing(label, class_num)
```
Comment (LDOUBLEV) [inline ppocr/losses/basic_loss.py:45]: label_smooth可以作为一个参数选项

File: ppocr/losses/basic_loss.py [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+    DMLLoss
+    """
+
+    def __init__(self, name="loss_dml"):
+        super().__init__()
+        self.name = name
+
+    def forward(self, out1, out2):
+        loss_dict = {}
+        soft_out1 = F.softmax(out1, axis=-1)
```
Comment (LDOUBLEV) [inline ppocr/losses/basic_loss.py:71]: 不一定是softmax，也可以是sigmoid，检测蒸馏里是sigmoid

File: ppocr/losses/basic_loss.py [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+    DistanceLoss:
+        mode: loss mode
+        name: loss key in the output dict
+    """
+
+    def __init__(self, mode="l2", name="loss_dist", **kargs):
+        assert mode in ["l1", "l2", "smooth_l1"]
+        if mode == "l1":
+            self.loss_func = nn.L1Loss(**kargs)
+        elif mode == "l1":
```
Comment (LDOUBLEV) [inline ppocr/losses/basic_loss.py:93]: mode == "l2"

File: ppocr/losses/basic_loss.py [hunk 8]
```diff
@@ -0,0 +1,101 @@
...
+        super().__init__()
+        self.name = name
+
+    def forward(self, out1, out2):
+        loss_dict = {}
+        soft_out1 = F.softmax(out1, axis=-1)
+        log_soft_out1 = paddle.log(soft_out1)
+        soft_out2 = F.softmax(out2, axis=-1)
+        log_soft_out2 = paddle.log(soft_out2)
+        loss = (F.kl_div(
```
Comment (LDOUBLEV) [inline ppocr/losses/basic_loss.py:75]: 这个函数的速度问题修复了吗？

File: ppocr/losses/distillation_loss.py [hunk 9]
```diff
@@ -0,0 +1,76 @@
...
+
+    def forward(self, predicts, batch):
+        loss_dict = dict()
+        for model_name in self.model_name_list:
+            out = predicts[model_name]
+            if self.key is not None:
+                out = out[self.key]
+            loss = super().forward(out, batch)
+            if isinstance(loss, dict):
+                assert len(loss) == 1
```
Comment (LDOUBLEV) [inline ppocr/losses/distillation_loss.py:73]: 这里为什么要验证==1，返回的loss可能是dict， 最终_loss=loss['loss']

File: ppocr/utils/save_load.py [hunk 10]
```diff
@@ -42,7 +42,10 @@ def _mkdir_if_not_exist(path, logger):
                 raise OSError('Failed to mkdir {}'.format(path))
 
 
-def load_dygraph_pretrain(model, logger, path=None, load_static_weights=False):
+def load_dygraph_pretrain(model,
+                          logger=None,
+                          path=None,
+                          load_static_weights=False):
```
Comment (WenmuZhou) [inline ppocr/utils/save_load.py:48]: 删掉load_static_weights参数吧，去掉name后，就不能从静态图加载参数了，可能config.md的文档里load_static_weights字段的说明也要删掉

File: configs/rec/ch_ppocr_v2.0/rec_chinese_lite_train_distillation_v2.1.yml (hunk 4 above)
Comment (littletomatodonkey) [reply configs/rec/ch_ppocr_v2.0/rec_chinese_lite_train_distillation_v2.1.yml]: 好的，已修改

File: configs/rec/ch_ppocr_v2.0/rec_chinese_lite_train_distillation_v2.1.yml (hunk 1 above)
Comment (littletomatodonkey) [reply configs/rec/ch_ppocr_v2.0/rec_chinese_lite_train_distillation_v2.1.yml]: 1. 好的，参数已经移动到子网络中去了，外面不会再暴露该接口
2. Global pretrain有的话，说明用户希望继续之前的训练过程，这块是需要覆盖的

File: configs/rec/ch_ppocr_v2.0/rec_chinese_lite_train_distillation_v2.1.yml (hunk 3 above)
Comment (littletomatodonkey) [reply configs/rec/ch_ppocr_v2.0/rec_chinese_lite_train_distillation_v2.1.yml]: 好的，已修改

File: configs/rec/ch_ppocr_v2.0/rec_chinese_lite_train_distillation_v2.1.yml (hunk 2 above)
Comment (littletomatodonkey) [reply configs/rec/ch_ppocr_v2.0/rec_chinese_lite_train_distillation_v2.1.yml:91]: 一般情况下，是只会处理特定子网络的输出，如果是模型融合的话，会用到多个模型的输出，这块需要重新写一个eval函数

File: ppocr/losses/basic_loss.py (hunk 5 above)
Comment (littletomatodonkey) [reply ppocr/losses/basic_loss.py:45]: 嗯嗯，现在的label smooth就是根据epsilon去判断的，不为None，就是label smooth

File: ppocr/losses/basic_loss.py (hunk 6 above)
Comment (littletomatodonkey) [reply ppocr/losses/basic_loss.py]: 好的哥哥，已经添加act参数

File: ppocr/losses/basic_loss.py (hunk 8 above)
Comment (littletomatodonkey) [reply ppocr/losses/basic_loss.py:85]: 不使用log_softmax api的话，速度是没有问题的

File: ppocr/losses/basic_loss.py (hunk 7 above)
Comment (littletomatodonkey) [reply ppocr/losses/basic_loss.py]: 好的，已修改

File: ppocr/utils/save_load.py (hunk 10 above)
Comment (littletomatodonkey) [reply ppocr/utils/save_load.py]: 好的，已删除

**Review by WenmuZhou — DISMISSED** [review]

File: configs/rec/ch_ppocr_v2.0/rec_chinese_lite_train_distillation_v2.1.yml [hunk 11]
```diff
@@ -0,0 +1,158 @@
...
+  debug: false
+  use_gpu: true
+  epoch_num: 800
+  log_smooth_window: 20
+  print_batch_step: 10
+  save_model_dir: ./output/rec_chinese_lite_distillation_v2.1
+  save_epoch_step: 3
+  eval_batch_step: [0, 2000]
+  cal_metric_during_train: true
+  pretrained_model: null
```
Comment (WenmuZhou) [inline configs/rec/ch_ppocr_v2.0/rec_chinese_lite_train_distillation_v2.1.yml:11]: 为null的直接留空白吧

File: ppocr/losses/basic_loss.py (hunk 5 above)
Comment (WenmuZhou) [reply ppocr/losses/basic_loss.py:44]: 两者的意思会不会不好联系在一起

File: ppocr/utils/save_load.py [hunk 12]
```diff
@@ -42,35 +42,10 @@ def _mkdir_if_not_exist(path, logger):
                 raise OSError('Failed to mkdir {}'.format(path))
 
 
-def load_dygraph_pretrain(model, logger, path=None, load_static_weights=False):
+def load_dygraph_pretrain(model, logger=None, path=None):
```
Comment (WenmuZhou) [inline ppocr/utils/save_load.py:45]: 这个函数这么短，要不直接删掉，相关代码放到init_model里吧

File: configs/rec/ch_ppocr_v2.0/rec_chinese_lite_train_distillation_v2.1.yml (hunk 11 above)
Comment (littletomatodonkey) [reply configs/rec/ch_ppocr_v2.0/rec_chinese_lite_train_distillation_v2.1.yml:11]: 好的，done

File: ppocr/utils/save_load.py (hunk 12 above)
Comment (littletomatodonkey) [reply ppocr/utils/save_load.py]: done, thx

**Review by LDOUBLEV — COMMENTED** [review]
补充个文档和config参数说明吧

File: ppocr/losses/basic_loss.py [hunk 13]
```diff
@@ -0,0 +1,103 @@
...
+
+    def forward(self, out1, out2):
+        if self.act is not None:
+            out1 = self.act(out1)
+            out2 = self.act(out2)
+
+        log_out1 = paddle.log(out1)
+        log_out2 = paddle.log(out2)
+        loss = (F.kl_div(
+            log_out1, out2, reduction='batchmean') + F.kl_div(
```
Comment (dyning) [inline ppocr/losses/basic_loss.py:81]: 为什么都不用log呢？

Comment (littletomatodonkey) [reply ppocr/losses/basic_loss.py:81]: 下面笔误，已更正，thx

**Review by littletomatodonkey — APPROVED** [review]

**Review by LDOUBLEV — COMMENTED** [review]
LGTM
